test/test_cpp_api_parity.py

```diff
@@ -0,0 +1,283 @@
+                    input_args=input_args)
+            for method in torch_nn_test_methods:
+                functions.append(test_params.module_variant_name + '_test_' + method)
+
+            # Just-in-time compile the C++ test code
+            cpp_module = torch.utils.cpp_extension.load_inline(
+                name=test_params.module_variant_name,
+                cpp_sources=cpp_source,
+                functions=functions,
+                verbose=False,
+            )
+            return cpp_module
+
+        def serialize_module_into_file(module, example_inputs):
+            # We use JIT tracing to serialize Python module state, so that we can load it into C++
+            traced_script_module = torch.jit.trace(module, example_inputs)
+            module_file = tempfile.NamedTemporaryFile(delete=False)
+            traced_script_module.save(module_file.name)
+            module_file.close()
+            return module_file.name
+
+        def test_method(method_name, test_params, cpp_module):
+            device = test_params.device
+            input_size = test_params.input_size
+            input_fn = test_params.input_fn
+            python_module_class = test_params.python_module_class
+            python_constructor_args = test_params.python_constructor_args
+            module_variant_name = test_params.module_variant_name
+            cpp_test_name = module_variant_name + '_test_' + method_name
+
+            if input_size:
+                example_inputs = [torch.randn(input_size)]
+            elif input_fn:
+                example_inputs = list(input_fn())
+            else:
+                raise RuntimeError("Missing `input_size` or `input_fn` for {}".format(module_variant_name))
+            example_inputs = [x.to(device) for x in example_inputs]
+
+            if method_name == 'init':
+                args = setup_init_test(device, python_module_class, python_constructor_args)
+            elif method_name == 'forward':
+                args = setup_forward_test(device, python_module_class, python_constructor_args, example_inputs)
+            elif method_name == 'backward':
+                args = setup_backward_test(device, python_module_class, python_constructor_args, example_inputs)
+            else:
+                raise RuntimeError("{} is not a supported method to test".format(method_name))
+
+            modules = args[0]
+            module_file_names = [serialize_module_into_file(module, example_inputs) for module in module]
+
+            cpp_args = module_file_names.copy()
+            for arg in args[1:]:
+                if isinstance(arg, list):
+                    cpp_args += arg
+                else:
+                    cpp_args.append(arg)
+            try:
+                cpp_test_fn = getattr(cpp_module, cpp_test_name)
+                if test_params.expect_parity_error:
+                    with self.assertRaisesRegex(RuntimeError, "Parity test failed"):
```

Comment:
I added this back because we have to check that the cpp test fails for all methods (init/forward/backward). If we use `unittest.expectedFailure` for this unit test, it will fail and return after the first method, while skipping the rest of the methods.

Reply:
Got it, that's reasonable.